src/Language/Haskell/GHC/ExactPrint.hs

```diff
@@ -144,47 +146,47 @@ setPos l = modify (\s -> s {epPos = l})
 
 -- ---------------------------------------------------------------------
 
--- |Given the original column when the
--- offset was calculated, a step offset to be applied, the column that
--- would have been used had the ruling offset been in play, and
--- whether we are moving onto a new line, determine an equivalent
+-- Given an annotation, determines a new offset relative to the previous
 -- offset
-pushOffset :: Annotation -> EP ()
-pushOffset (Ann (DP (edl,edc)) nl sc dc _kds) = do
-  (co,cd) <- gets (ghead "pushOffset" . epStack)
-  (_l,c) <- getPos
-
-  epStack' <- gets epStack
+--
+withOffset :: Annotation -> (EP () -> EP ())
+withOffset a@(Ann (DP (edLine, edColumn)) newline originalStartCol annDelta _) k = do
+  (colOffset, colDelta) <- asks epStack
+  (_l, currentColumn) <- getPos
   let
-      (co'',cd') = case nl of
-                    KeepOffset  -> (dc + cd,cd)
-                    LineChanged -> (dc + cd,cd)
-                    LineSame ->
-                      let
-                        co' = dc + co
-                        nd = - (sc - oc)
-                        oc = if edl == 0 then  c + edc -- same line
-                                         else co + edc -- different line
-                      in
-                        (co' - (cd - nd), nd)
-  modify (\s -> s {epStack = (co'',cd'): epStack s})
-    `debug` ("pushOffset:((edl,edc),nl,sc,dc,(co,cd),c,(co'',cd'),epStack')="
-                 ++ show ((edl,edc),nl,sc,dc,(co,cd),c,(co'',cd'),epStack'))
+      newOffset =
+        case newline of
+          LineChanged -> (annDelta + colDelta , colDelta)
+          KeepOffset -> (annDelta + colDelta , colDelta)
+          LineSame ->
+            let
+              -- Add extra indentation
+              colOffset' = annDelta + colOffset
+              -- Work out where new column starts
+              -- If the annLineDelta == 0 ie, we stay on the same
+              -- line so the new column is the current position
+              --
+              --
+              -- If not then we move to the next line so the
+              -- start column is whatever the indentation context
+              -- is `colOffset`
+              newStartColumn = edColumn +
+                                 if edLine == 0
+                                   then currentColumn -- same line
+                                   else colOffset -- different line
+              newColDelta = newStartColumn - originalStartCol
+              in ((colOffset' + newColDelta) - colDelta, newColDelta)
```

Comment:
The original made it clear that we are adjusting by the difference between cd and nd. Perhaps change this to be

    colOffset' + (newColDelta - colDelta)

to emphasis this.

There are essentially two different things happening, the annDelta is coming in, and the current colDelta is being changed, but only if it is different now from what it was before.